discovery/aws/ec2.go

```diff
@@ -233,14 +244,14 @@ func (d *EC2Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error
 	}
 
 	input := &ec2.DescribeInstancesInput{Filters: filters}
-
+	azIDsRefreshAttempted := false
```

Comment:
I would just call refresh AZ ID's here then, since we only want to call it once anyway. therefore we can use a bool instead of a pointer to a bool.

Reply:
I originally wrote it this way as it's very unlikely a refresh will be needed - only if AWS adds a new AZ to a region! This saves a request every minute by every Prometheus process using EC2 SD.

Reply:
But the new code does not refresh anymore if I am correct.

I think it would be fine to "reload" prometheus config (that restarts the service discoveries) if aws adds a new AZ to a region and you happen to have machines here.

Reply:
> But the new code does not refresh anymore if I am correct.

The new code retrieves the list of AZ IDs if it doesn't recognise an instance's AZ, *provided it hasn't already done so during the current `refresh()` call*.

> I think it would be fine to "reload" prometheus config (that restarts the service discoveries) if aws adds a new AZ to a region and you happen to have machines here.

This is how the first version of the original PR worked, however you [requested](https://github.com/prometheus/prometheus/pull/8896#issuecomment-855474667) dynamic handling of new AZs, which was fair. If we attempted this request only once when the SD starts up, what's the best way to handle a failure?

Reply:
Okay so this is per-refresh.

In general given the fact that existing tokens might not have the right to list the AZ's, it would be acceptable to try to refresh the AZ's once every refresh, if then len() of AZ ID's == 0 (so we do not need the booleans).

I think we could log the failure as Debug if the failure is because of permission issues, otherwise Warn.

See https://github.com/prometheus/prometheus/blob/main/discovery/aws/ec2.go#L334 for how we could filter the errors.

Reply:
Ok that makes sense. One last question: if, despite our best efforts, we can't retrieve the AZ ID, do you think a warning/debug message should be printed for every target without a match (as it currently is), or should the label be silently set to the empty string?

Reply:
I've updated this as agreed; the only thing I've changed is to log a warning rather than debug message if the `DescribeAvailabilityZones` call fails with an auth error. Not having `ec2:DescribeInstances` is always logged with error severity, so it seems consistent to log a warning if discovery is partially successful. My concern with debug is that it is completely silent at the default log level, and means folks would have to check the docs to realise the AZ ID label is actually available, and even then they might wonder why it's not appearing. Only those already using EC2 SD would see this warning, so I think it's worth having - what do you think? If a given AZ ID is not found, that is a debug message, and the `*bool` is also gone.